src/rest/RequestHandler.js

```diff
@@ -0,0 +1,177 @@
+const DiscordAPIError = require('./DiscordAPIError');
+const HTTPError = require('./HTTPError');
+const Util = require('../util/Util');
+const { Events: { RATE_LIMIT }, browser } = require('../util/Constants');
+
+function parseResponse(res) {
+  if (res.headers.get('content-type').startsWith('application/json')) return res.json();
+  if (browser) return res.blob();
+  return res.buffer();
+}
+
+class RequestHandler {
+  constructor(manager) {
+    this.manager = manager;
+    this.busy = false;
+    this.queue = [];
+    this.reset = -1;
+    this.remaining = -1;
+    this.limit = -1;
+    this.retryAfter = -1;
+  }
+
+  push(request) {
+    this.queue.push(request);
+    this.run();
+  }
+
+  run() {
+    if (this.queue.length === 0) return;
+    this.execute(this.queue.shift());
+  }
+
+  get limited() {
+    return (this.manager.globallyRateLimited || this.remaining <= 0) && Date.now() < this.reset;
+  }
+
+  get _inactive() {
+    return this.queue.length === 0 && !this.limited && this.busy !== true;
+  }
+
+  _getAPIOffset(serverDate) {
+    return new Date(serverDate).getTime() - Date.now();
+  }
+
+  _calculateReset(reset, serverDate) {
+    return new Date(Number(reset) * 1000).getTime() - this._getAPIOffset(serverDate);
+  }
+
+  /* eslint-disable-next-line complexity */
+  async execute(item) {
+    // Insert item back to the beginning if currently busy
```

Comment:
you may be able to save some code by destructuring `item` (outside of the parameter definition ofc since you do reuse the item)